use odom gearset/drivescales when available

dcieslak19973 opened the pull request:
### Description of the Change

When using geared drivetrains and odom, the odom drivescales are not used, leading to inaccurate movements

### Motivation

Correct behavior described above

### Possible Drawbacks

N/A

### Verification Process

Physical test

### Applicable Issues

Fixes #462

dcieslak19973 committed:
use odom gearset/drivescales when available

theol0403 commented:
Fyi: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request

Octogonapus commented on src/impl/chassis/controller/chassisControllerBuilder.cpp:
```diff
@@ -413,6 +413,12 @@ ChassisControllerBuilder::buildDOCC(std::shared_ptr<ChassisController> chassisCo
+  auto ccGearset = gearset;
+  auto ccDriveScales = driveScales;
+  if(differentOdomScales) {
+    ccDriveScales = odomScales;
```
Can't we just use `odomScales` in either case?

dcieslak19973 replied:
Will test whether that works.

Octogonapus commented on src/impl/chassis/controller/chassisControllerBuilder.cpp:
```diff
@@ -413,6 +413,12 @@ ChassisControllerBuilder::buildDOCC(std::shared_ptr<ChassisController> chassisCo
+  auto ccDriveScales = driveScales;
+  if(differentOdomScales) {
+    ccDriveScales = odomScales;
+    ccGearset = AbstractMotor::gearset::green;
```
Why hard-code green?

dcieslak19973 replied:
Yeah, that may not be necessary. will test

dcieslak19973 replied:
@Octogonapus - During physical testing I ran into issues with an externally geared chassis which I've also resolved in this PR. Feedback?

dcieslak19973 committed:
Simplified change

dcieslak19973 committed:
Additional fixes for externally geared drive trains

Ran into issues testing on a physical robot that has an externally geared drivetrain. This produces correct behavior

Octogonapus commented on src/api/chassis/controller/chassisControllerPid.cpp:
```diff
@@ -126,7 +127,7 @@ void ChassisControllerPID::moveDistanceAsync(const QLength itarget) {
   mode = distance;
 
-  const double newTarget = itarget.convert(meter) * scales.straight * gearsetRatioPair.ratio;
+  const double newTarget = itarget.convert(meter) * scales.straight;
```
Is this change (and the one below) backwards compatible for people that don't call `withDimensions` with a `GearsetRatioPair`?

dcieslak19973 replied:
Yes, if one does `.withDimensions(AbstractMotor::gearset::green, {{4.0_in,10_in}, imev5GreenTPR})` it will create with a gearsetRatioPair with a 1.0 gear ratio defaulted.

Octogonapus replied:
I am not sure where the external ratio is accounted for, then. It seems like it is always discarded.

dcieslak19973 replied:
@Octogonapus - you are correct. Latest commit should fix this. The chassisControllerPID now works physically with both external tracking wheels and configured to use the drive train as the odom source.

I've phyically tested this with both configurations

dcieslak19973 committed:
Fixed CCPID issue with integrated odom vs tracking wheel

dcieslak19973 committed:
Updated documentation

Octogonapus reviewed (approved):
Thanks for working on this rats nest :+1:
